### PR TITLE
Add an X parameter to the Likelihood.

### PR DESCRIPTION
**Pull request by jesnie**
Add an `X` parameter to the `Likelihood` class methods.

This will make it easy/possibly to write heteroskedastic likelihoods.

I plan to do more with heteroskedastic likelihoods in downstream PRs, but I want to get the "controversial" part of it out of the way first.

Basically this is the controversial part, because this is a breaking change.
To guesstimate the scale of the breakage, consider our notebooks. We have 35 notebooks doing all kinds of advanced stuff. A single one of those breaks.

### CODE REVIEW

File: gpflow/likelihoods/multiclass.py [hunk 1]
```diff
@@ -34,17 +34,17 @@ class Softmax(MonteCarloLikelihood):
     """
 
     def __init__(self, num_classes: int, **kwargs: Any) -> None:
-        super().__init__(latent_dim=num_classes, observation_dim=None, **kwargs)
+        super().__init__(input_dim=None, latent_dim=num_classes, observation_dim=None, **kwargs)
         self.num_classes = self.latent_dim
 
-    def _log_prob(self, F: TensorType, Y: TensorType) -> tf.Tensor:
+    def _log_prob(self, X: TensorType, F: TensorType, Y: TensorType) -> tf.Tensor:
```
Comment (vdutor) [inline gpflow/likelihoods/multiclass.py:40]: Could you explain why you choose this order of inputs?

File: gpflow/likelihoods/base.py [hunk 2]
```diff
@@ -48,97 +50,119 @@ def __init__(self, latent_dim: Optional[int], observation_dim: Optional[int]) ->
         :param observation_dim: the dimension of the observation vector Y for a single data point
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:50]: Forgot to add `input_dim` to docstrings

File: gpflow/likelihoods/base.py [hunk 3]
```diff
@@ -48,97 +50,119 @@ def __init__(self, latent_dim: Optional[int], observation_dim: Optional[int]) ->
...
-        :param Y: observation Tensor, with shape [..., observation_dim]
+        :param result: result tensor, with shape [...]
+        :param X: input tensor, with shape [..., input_dim]
+        :param F: function evaluation tensor, with shape [..., latent_dim]
+        :param Y: observation tensor, with shape [..., observation_dim]
         """
         expected_shape = tf.broadcast_dynamic_shape(tf.shape(F)[:-1], tf.shape(Y)[:-1])
         tf.debugging.assert_equal(tf.shape(result), expected_shape)
 
+    def _check_input_dims(self, X: TensorType) -> None:
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:84]: I assume you will be adding your shape-checking functionality in a later PR?

File: gpflow/likelihoods/base.py [hunk 4]
```diff
@@ -274,7 +313,7 @@ def _quadrature_dim(self) -> int:
         assert self.latent_dim is not None
         return self.latent_dim
 
-    def _quadrature_log_prob(self, F: TensorType, Y: TensorType) -> tf.Tensor:
+    def _quadrature_log_prob(self, F: TensorType, X: TensorType, Y: TensorType) -> tf.Tensor:
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:316]: why this change in order? Is there a way of working around this?

File: gpflow/likelihoods/base.py [hunk 5]
```diff
@@ -48,97 +50,119 @@ def __init__(self, latent_dim: Optional[int], observation_dim: Optional[int]) ->
         :param observation_dim: the dimension of the observation vector Y for a single data point
         """
         super().__init__()
+        self.input_dim = input_dim
         self.latent_dim = latent_dim
         self.observation_dim = observation_dim
 
-    def _check_last_dims_valid(self, F: TensorType, Y: TensorType) -> None:
+    def _check_last_dims_valid(self, X: TensorType, F: TensorType, Y: TensorType) -> None:
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:57]: Can you explain why you choose this order for the input args?

**Review by vdutor — COMMENTED** [review]
In principle, I am happy with this proposal. The only thing is that sometimes the order of input args seems inconsistent. It would be good to know why that is the case.

File: gpflow/likelihoods/base.py (hunk 5 above)
Comment (jesnie) [reply gpflow/likelihoods/base.py:58]: > Can you explain why you choose this order for the input args?

It seems most intuitive to me. You start with an X, then you map that to F, then you map that to Y?

File: gpflow/likelihoods/base.py (hunk 2 above)
Comment (jesnie) [reply gpflow/likelihoods/base.py:51]: > Forgot to add `input_dim` to docstrings

Done.

File: gpflow/likelihoods/base.py (hunk 3 above)
Comment (jesnie) [reply gpflow/likelihoods/base.py:85]: > I assume you will be adding your shape-checking functionality in a later PR?

Yes.

File: gpflow/likelihoods/base.py (hunk 4 above)
Comment (jesnie) [reply gpflow/likelihoods/base.py:317]: > why this change in order? Is there a way of working around this?

This is because this method is used as `self.quadrature.logspace(self._quadrature_log_prop, ...)` and `self.quadrature` operates over the first parameter, so the first parameter has to be `F`.

To work around this we'd have to update all our quadrature stuff to have special-case handling of `X`, which seems like a bad idea.

**Comment by jesnie** [comment]
> In principle, I am happy with this proposal. The only thing is that sometimes the order of input args seems inconsistent. It would be good to know why that is the case.

Basically, our quadrature code operates over the first argument, so a couple of places I have had to change parameter order to have `F` as the first argument.

File: RELEASE.md [hunk 6]
```diff
@@ -39,6 +39,8 @@ This release contains contributions from:
 
 ## Breaking Changes
 
+* All likelihood methods now take an extra `X` parameter. If you have written custom likelihoods or
+  you have custom code calling likelihoods directly you will need to add this extra parameter.
```
Comment (frgsimpson) [inline RELEASE.md:43]:
```suggestion
* All likelihood methods now take an extra `X` argument. If you have written custom likelihoods or
  you have custom code calling likelihoods directly you will need to add this extra argument.
```
Just to avoid confusion with the gpflow Parameter object.

File: RELEASE.md [hunk 7]
```diff
@@ -50,6 +52,9 @@ This release contains contributions from:
 
 ## Major Features and Improvements
 
+* Improved handling of variable noise
+  - All likelihood methods now take an `X` parameter, allowing you to easily implement
```
Comment (frgsimpson) [inline RELEASE.md:56]:
```suggestion
  - All likelihood methods now take an `X` argument, allowing you to easily implement
```

**Review by frgsimpson — APPROVED** [review]
Looks great!

File: RELEASE.md (hunk 6 above)
Comment (jesnie) [reply RELEASE.md:43]: > Just to avoid confusion with the gpflow Parameter object.

Sure.

File: RELEASE.md (hunk 7 above)
Comment (jesnie) [reply RELEASE.md:56]: Sure. done.
